torch/csrc/jit/export.cpp

```diff
@@ -638,6 +642,186 @@ void ScriptModuleSerializer::writeTensorTable(torch::ModelDef* model_def) {
+    // module def
+    push(OpCode::EMPTY_LIST);
+    push(OpCode::MARK);
+  }
+  void finish() {
+    push(OpCode::APPENDS);
+    push(OpCode::STOP);
+  }
+
+  uint64_t addIValue(const IValue& ivalue) {
+    if (ivalue.isTensor()) {
+      // Write it to the tensor table
+      push(OpCode::EXT1);
+      push(ExtensionCode::TENSOR);
+      auto tensor_id = serializer_.addTensor(ivalue.toTensor());
+      push(OpCode::BININT, uint32_t(tensor_id));
+    } else if (ivalue.isBlob()) {
+      AT_ERROR("Unsupported IValue type for pickling (Blob)");
+    } else if (ivalue.isTuple()) {
+      pushTuple(ivalue);
+    } else if (ivalue.isDouble()) {
+      pushDouble(ivalue);
+    } else if (ivalue.isFuture()) {
+      AT_ERROR("Unsupported IValue type for pickling (Future)");
+    } else if (ivalue.isInt()) {
+      // TODO: use BININT1/BININT2/LONG if possible/necessary
+      push(OpCode::BININT, uint32_t(ivalue.toInt()));
+    } else if (ivalue.isBool()) {
+      if (ivalue.toBool()) {
+        push(OpCode::NEWTRUE);
+      } else {
+        push(OpCode::NEWFALSE);
+      }
+    } else if (ivalue.isIntList()) {
+      AT_ERROR("Unsupported IValue type for pickling (IntList)");
+    } else if (ivalue.isString()) {
+      pushString(ivalue);
+    } else if (ivalue.isDoubleList()) {
+      AT_ERROR("Unsupported IValue type for pickling (DoubleList)");
+    } else if (ivalue.isBoolList()) {
+      AT_ERROR("Unsupported IValue type for pickling (BoolList)");
+    } else if (ivalue.isTensorList()) {
+      AT_ERROR("Unsupported IValue type for pickling (TensorList)");
+    } else if (ivalue.isGenericList()) {
+      pushList(ivalue);
+    } else if (ivalue.isGenericDict()) {
+      pushDict(ivalue);
+    } else if (ivalue.isObject()) {
+      AT_ERROR("Unsupported IValue type for pickling (Object)");
+    } else if (ivalue.isNone()) {
+      push(OpCode::NONE);
+    } else if (ivalue.isDevice()) {
+      AT_ERROR("Unsupported IValue type for pickling (Device)");
+    } else {
+      AT_ERROR("Unknown IValue type for pickling");
+    }
+
+    return 0;
+  }
+
+private:
+  void pushString(const IValue& ivalue) {
+    auto string = ivalue.toStringRef();
+
+    push(OpCode::BINUNICODE);
+    push(uint32_t(string.size()));
+    stack_.insert(stack_.end(), string.begin(), string.end());
+
+    push(OpCode::BINPUT, memo_id++);
+  }
+
+  void pushDouble(const IValue& ivalue) {
+    double value = ivalue.toDouble();
+    AT_ASSERT(sizeof(double) == 8);
+    char bytes[8];
+    double* bytes_as_double = reinterpret_cast<double*>(bytes);
+    *bytes_as_double = value;
+
+    push(OpCode::BINFLOAT);
+    for (size_t i = 0; i < sizeof(bytes); ++i) {
+      push(bytes[sizeof(bytes) - i - 1]);
+    }
+  }
+
+  void pushDict(const IValue& ivalue) {
+    auto dict = ivalue.toGenericDictRef();
+
+    push(OpCode::EMPTY_DICT);
+    push(OpCode::BINPUT, memo_id++);
+
+    push(OpCode::MARK);
+
+    for (auto pair : dict) {
```

Comment:
Can we impose an order on the dictionary elements? It would be really sad if saving a dictionary was nondeterministic. Would make reproducible models impossible. (not needed for this diff, but is something we should not forget to do before the next release)

Reply:
That would be difficult since right now dictionaries are backed by `std::unordered_map`. This may also be worth revisiting since dicts are insertion ordered in Python 3.7+.

Reply:
<in person> Use sorting to get deterministic order (only valid keys are int, float, str at the moment).